parsec/mca/device/device_gpu.c

```diff
@@ -1660,6 +1667,48 @@ parsec_device_callback_complete_push(parsec_device_gpu_module_t   *gpu_device,
             parsec_data_end_transfer_ownership_to_copy(task->data[i].data_out->original,
                                                        gpu_device->super.device_index,
                                                        flow->flow_flags);
+
+            parsec_data_copy_t* source = gtask->sources[i];
+            parsec_device_gpu_module_t *src_device =
+                    (parsec_device_gpu_module_t*)parsec_mca_device_get( source->device_index );
+            if (task->data[i].data_in->flags & PARSEC_DATA_FLAG_EVICTED) {
+                /**
+                 * The device copy had been evicted to the host and brought back in.
+                 * If this is the only device on which that data is used we can release
+                 * the host memory back to the application. If there are other devices
+                 * we cannot release the host memory because the data may actually be used
+                 * by a host task (e.g., after being sent there from a different device)
+                 * or be used as input to the other device.
+                 */
+                parsec_data_copy_t *cpu_copy = task->data[i].data_out->original->device_copies[0];
+                parsec_data_copy_t *gpu_copy = task->data[i].data_out;
+                parsec_data_t *original = task->data[i].data_out->original;
+                /* release host memory if requested */
+                if (cpu_copy->device_private != NULL &&
```

Comment:
A copy used as input existed before but has no reserved space on the device right now because it has been evicted. This means that at some point a predecessor task was located on the device, had memory for this copy, but that memory was since repurposed and the data relocated on the main memory. But the copy remained valid for some reason that I can't just figure out. Does this means PaRSEC can now propagate device copies ro successors instead of CPU copies and these device copies will survive even when evicted from the device ?

If that's the case it sounds like a good idea, but I wonder how is the thread-safety of these reuse vs. evictions handled ? How are the reference counts of the device copies managed (a ref count of an input is released at the end of the task), but in this case which copy shall be released, the GPU that existed but had no memory or the CPU that was not supposed to be referenced but existed as a placeholder for the data?

Reply:
The main use case here is for data that is exclusively produced on the device and moved between devices. The only time we need to allocate host memory for it is when the data is evicted. With the API as it is today, we have to pass a `parsec_data_copy_t` for the input flow. We use the host copy, with everything set but the `device_private`. That way we describe what the data should look like (size and type) without actually allocating memory.

Nothing else changes. Nothing else has to change. Device copies are not affected by this change.